### ISSUE TITLE
Create a c10::complex

### ISSUE THREAD

**Issue by zasdfgbnm** [pytorch/pytorch#35284]
**For the list of PRs, see https://github.com/pytorch/pytorch/issues/35284#issuecomment-605350370**

It is very unfortunate that, `std::complex` does not work on CUDA, so on CUDA, all complex operators are implemented as converting `std::complex` to `thrust::complex` and then work on that. But this approach has its issue too:

First, `thrust::complex` itself has some problem, for example, on CUDA9 and ROCm, it does not have a default constructor, which makes it very annoying when working with complex, and it could take a long time to debug such build errors. For example, there is a `ATen/native/cuda/CUDA9Workarounds.cuh` created to workaround this problem.

Second, this makes the dtype<-->scalar_t diverge on CPU and CUDA, this is really bad, because there are helpers in `c10/core/ScalarType.h` and `c10/util/TypeCast.h` that makes use of this assumption, and I believe most of our engineers also have this assumption in mind. Forgetting this fact could lead to subtle bugs, for example, I just noticed that the `needs_dynamic_casting` in `ATen/native/cuda/Loops.cuh` always returns true when the dtype is complex, because `thrust::complex` and `std::complex` are not the same type.

Such kind of behavior could lead to subtle bugs, for example: the copy kernel can not convert complex: https://github.com/pytorch/pytorch/issues/35225, also this PR https://github.com/pytorch/pytorch/pull/34749 also fails on complex, but only on Linux (Windows is fine). The issue and PR I just mentioned fails with `unspecified launch failure`. Although I have not been able to find the exact problem of the issue and PR I've just mentioned, the behavior of https://github.com/pytorch/pytorch/pull/34749 makes me feel like the issue is on CUDA itself instead of PyTorch.

Considering that we already have a `at::ComplexHalf`, I don't think having a `c10::complex` is a bad idea. And by creating our own implementation of complex at c10, we will have more control on it, hopefully it will solve these issues more cleanly.

It is not hard to implement `c10::complex`, for arithmetic and `real`, `imag`, it's trivial. For functions like `exp`, `log`, `sin`, it can be simply implemented as casting to `std::complex` or `thrust::complex` and then call the library.

After creating a `c10::complex`, all the `std::complex` in `c10/core/ScalarType.h` will be replaced with `c10::complex`.

The only problem that I can think of is backward compatibility. If some user has write something like

```C++
template<typename T>
struct is_complex: public false_type {};
template<typename T>
struct is_complex<std::complex<T>> : public true_type {};
```

it will not work.

cc @ezyang @anjali411 @dylanbespalko

**Comment by zasdfgbnm** [comment]
cc: @ngimel

**Comment by ezyang** [comment]
Don't worry about BC. We can introduce our own complex type. However, I think it would be clearer for the discussion if you could write a sketch of what the class would look like (including implicit conversions, etc) and how you are planning to deal with `thrust::complex` operations. Are you going to basically do "thrust::complex but with a default constructor"?

**Comment by dylanbespalko** [comment]
I think this a really good idea. I think the current solution to cast to `thrust::complex` or `rocm::complex` is a problem especially when you try to go beyond simple point-wise operators.

I have encountered code on the CPU and FPGA that tries to support BOTH "riri" (float[2]) and "rrii" memory patterns. This code pre-dates wide memory lanes which are now common on the CPU, GPU, and FPGA (virtually all processors!!!). We should not worry about BC and we should just define a c10::complex type that implies a float[2] memory layout.

**Comment by zasdfgbnm** [comment]
@ezyang Sure, I will prepare a sketch

**Comment by zasdfgbnm** [comment]
The rough idea is:
- `c10::complex` will be a complete implementation of complex type, the API of it will be mostly the same as `std::complex` in C++20.
- Member functions and operators are implemented inside c10.
- Math functions in namespace `std::` and `::` are overloaded, i.e. there will be something like `std::exp(c10::complex)`, the list of overloaded functions follows the C++ standard of `std::complex`, which can be found at https://en.cppreference.com/w/cpp/numeric/complex
- We don't implement math functions by ourselves, the implementation of math function will be something like:
  ```C++
  namespace std {

  C10_HOST_DEVICE c10::complex<float> exp(c10::complex<float> x) {
  #if defined(__CUDACC__) || defined(__HIPCC__)
    auto xx = static_cast<thrust::complex<float>>(x);
    auto yy = thrust::exp(xx);
    return static_cast<c10::complex<float>>(yy);
  #else
    auto xx = static_cast<std::complex<float>>(x);
    auto yy = std::exp(xx);
    return static_cast<c10::complex<float>>(yy);
  #endif
  }

  }  // namespace std
  ```
- ^^^ The above three features of `c10::complex` makes it a self-contained class, users could use arithmetics and math functions as if there were using `std::complex`, and the result would also be a `c10::complex`
- There will be converting constructor, assignment operators, and casting operators for `std::complex` and also `thrust::complex` as well on CUDA. Casting between `c10::complex` and `std::complex`/`thrust::complex` are all explicit. The behavior of being explicit makes it clear that which overload is called when applying math functions. If it was implicit, it might silently cast to either `std::complex` which is not defined on GPU, or there might be a collision on which implicit casting to choose (is it `std::complex`, or `thrust::complex`?).

**Comment by dylanbespalko** [comment]
This looks good. Not sure about the `namespace std { ... }` wrapping. Wouldn't this be in the `c10` namespace?

This could also serve as a initial step towards adding a `ComplexPolar` dtype that stores Mag/Arg instead of Re/Im. Looks great.

**Comment by zasdfgbnm** [comment]
@dylanbespalko For example, if I want to implement `exp`, it's likely I will write something like

```C++
  AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND1(ScalarType::Half, iter.dtype(), "exp_cuda", [&]() {
    gpu_kernel(iter, []GPU_LAMBDA(scalar_t a) -> scalar_t {
      return std::exp(a);
    });
  });
```

If I change this to

```C++
  AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND1(ScalarType::Half, iter.dtype(), "exp_cuda", [&]() {
    gpu_kernel(iter, []GPU_LAMBDA(scalar_t a) -> scalar_t {
      return c10::exp(a);
    });
  });
```

Then I need to create an implementation for all math functions with all dtypes in namespace `c10`, this will be lots of work.

**Comment by dylanbespalko** [comment]
@zasdfgbnm 

Ok, sorry, It looked weird to me.

**Comment by ezyang** [comment]
OK, this SGTM.

**Comment by zasdfgbnm** [comment]
The current plan is, I first prototype it on a standalone repository https://github.com/zasdfgbnm/c10-complex, once it's working, I will copy-paste it to PyTorch and submit a PR with just the implementation of the `c10:complex`, nothing will be using `c10:complex` at this PR. This worth a separate PR because, as what I have experienced with implementing it, different versions of NVCC, ROCm, Visual Studio, different compilers have different bugs, it's actually not trivial to implement such a thing that builds on all platforms we support.

Once the first PR is merged, In the second PR I will change the macros in `c10/core/ScalarType.h` to use `c10:complex<float>` and `c10:complex<double>`, and change a minimum amount of codes that uses complex to make PyTorch build and pass tests.

Then I am planning to run a `grep -i complex` on the whole code, and send small patches to fix all the uses.

**Comment by zasdfgbnm** [comment]
List of things:
- [ ] Core class (#35524)
- [ ] Math functions
- [ ] Make `c10::ComplexHalf` a template specialization of `c10::complex` (#35583)

**Comment by mruberry** [comment]
Separately, can we ask NVIDIA to implement std::complex in a future CUDA? We'll still need c10::Complex in the meanwhile.

**Comment by ezyang** [comment]
when someone gets around to doing this please post the nvbug here

**Comment by zasdfgbnm** [comment]
@ezyang The nvbug is http://nvbugs/2914116, but it is not customer viewable.

There were some discussions about the expected time, prioritized list of functions, etc. between them and me. If you want, you can open a new issue, and make discussions there, so that Facebook can provide a more accurate requirement description based on your internal plan for complex tensors.

**Comment by ezyang** [comment]
Thanks! I think you actually have most of the context here: we're trying to add complex number support, and we need to make a choice what `scalar_t` we will use in this case. We don't have FB prod use cases for complex right now; @anjali411 knows more about what researcher use cases we are targeting at the moment.

**Comment by zasdfgbnm** [comment]
Dispatch macros migration status:
| Macro                                                | Status  |
|------------------------------------------------------|---------|
| `AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES`       | 15 left |
| `AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND1` | 17 left |
| `AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND2` | 9 left  |
| `AT_DISPATCH_COMPLEX_TYPES`                      | 8 left  |
| `AT_DISPATCH_ALL_TYPES_AND_COMPLEX`            | Done    |
| `AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND`       | Done    |
| `AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND2`      | Done    |
| `AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3`      | Done    |

**Comment by zasdfgbnm** [comment]
c10::complex is done

**Comment by kvathupo** [comment]
> Separately, can we ask NVIDIA to implement std::complex in a future CUDA? We'll still need c10::Complex in the meanwhile.

Apologies in advance for unearthing an old thread, but `std::complex` like functionality seems to have been added in CUDA 11.3 with `cuda::std::complex` from libcu++ (also open-source) [1]. Pytorch doesn't support that CUDA version yet, but it's worth keeping in mind for future `c10::complex` development. 

Perhaps you can replace `c10::complex` stuff with libcu++, depending on if c10 supports things `cuda::std::complex` doesn't atm, e.g. `long double`.


[1] - https://nvidia.github.io/libcudacxx/releases.html

**Comment by anjali411** [comment]
@kvathupo thanks for sharing that! we do support CUDA 11.3. In fact, we also support 11.6 and 11.7, so I think it might be a good idea to switch. Currently, we c10::complex calls into thrust, and IIRC, we have had some consistency issues between the thrust and standard library, so it might be a good idea to switch to this. cc. @ngimel

**Comment by ngimel** [comment]
cc @zasdfgbnm

**Comment by zasdfgbnm** [comment]
> and IIRC, we have had some consistency issues between the thrust and standard library, so it might be a good idea to switch to this.
I think thrust is doing a better job than the standard library (LLVM's libc++ and libcu++). That's why we haven't switched.

**Comment by anjali411** [comment]
> I think thrust is doing a better job than the standard library (LLVM's libc++ and libcu++)

Ha in what way specifically? It might be useful to document this somewhere for anyone who looks at this in the future.

**Comment by zasdfgbnm** [comment]
> Ha in what way specifically?
I didn't check the performance, but in terms of numerical stability, thrust does better.
